### ISSUE TITLE
difficulty in reproducing your result

### ISSUE THREAD

**Issue by chuanzihe** [e-lab/ENet-training#15]
Hi, 

Your team mentioned the significance of setting batch size in training. Then may I know how do you explain why batch size would impact on the final result so strongly? 

I have trained with a batch size of 2 and adjusted my learning rate to 1e-5. I also modified the original code by adding a  **iterSize** of 4. In essence, the real batch size is **8**. However, it wouldn't achieve the performance of the pre-trained model you provided.

Here's the error curve: 
**Encoder**
<img width="553" alt="screen shot 2016-09-10 at 4 18 09 pm" src="https://cloud.githubusercontent.com/assets/19618625/18409117/f82bdb4c-7772-11e6-871b-e10689e4757b.png">

**Decoder**
<img width="549" alt="screen shot 2016-09-10 at 4 18 20 pm" src="https://cloud.githubusercontent.com/assets/19618625/18409120/00946844-7773-11e6-8974-df2283b4d641.png">

My result:

![lr5](https://cloud.githubusercontent.com/assets/19618625/18409179/0542ed9c-7774-11e6-88bc-1315ca9c62cb.png)

Thank you for answering.

**Comment by xmba15** [comment]
Hi, have you made any improvements now?
I am trying to implement ENet on top of caffe too and test on City Scape dataset whose images are resized to the size of  512x512. I cannot get the accuracy proposed in the paper. Any suggestions on weight initialization when implementing the network on Caffe?

**Comment by chuanzihe** [comment]
@bajsk Hi, I have not tried it on Caffe. BTW how did you convert the model to Caffe model? How did you handle the parameters in BN layers?

**Comment by codeAC29** [comment]
@clairematrix what is the image size which you are using?
@bajsk Sorry I won't be able to help you in Caffe. Btw we used image size of 256x512 and you cannot use a resolution of 512x512.

**Comment by chuanzihe** [comment]
@codeAC29 Hi, I loaded the original Cityscape image: 2048 × 1024. When training, I followed your README.MD. 
| [Cityscapes](https://www.cityscapes-dataset.com/) | 512x256 | 64x32 | 19 |

**Comment by xmba15** [comment]
@clairematrix  I tried to implement it using layers on caffe. I used BN layer on caffe without a Scale layer and set eps to 1e-3 as in Torch implementation.
@codeAC29 Can you tell me why we cannot use this network for resolution of 512x512? Actually, I resized the original images and labelled training data to 512x512 and tried to construct the network as described in your paper. You described an example with the input of 512x512 in Table 1 in the paper, right?

**Comment by codeAC29** [comment]
@clairematrix The result which you have posted does not look like a 512x256 image. Ideally you should use the same resolution for both training and testing.
@bajsk The original image is of the size 1024x2048. If you rescale it to 512x512 then you are changing the aspect ratio. In this case the network will be learning, for example a squeezed version of car or person, which you do not want to.

**Comment by ramonss** [comment]
Could someone share the full list of options settings that were used to achieve the results for CityScapes in the paper? I'm also having some difficulties to reproduce the result. Currently, I'm using 4x Titan X GPUs with the following command:

--learningRate 5e-4 --weightDecay 2e-4 --batchSize 10 --nGPU 4 --maxepoch 1000

I tried to follow the suggestion from the paper in terms of parameters for learning rate and weight decay. But how about the total number of epochs or other settings? I'm getting results similar to the following screenshot. In general it shows problems for detecting the road near the ego vehicle. Is that expected?

![image](https://cloud.githubusercontent.com/assets/5029025/19241116/69ac2892-8f0d-11e6-854d-6b04a5ecedef.png)

Do I need to use some special parameters to visualize the results? This is the command I used:

qlua demo.lua -d train/trained_models/ --net 320 -i train/data/Cityscapes/leftImg8bit/test/bonn/ -m cityscapes

The net number used above is the best epoch from the decoder training so far after 350 epochs of decoder training and 1000 epochs of encoder training.

I appreciate any help in advance.

**Comment by chuanzihe** [comment]
@codeAC29 Thanks. I do use the same resolution for both training and testing. The image I post was rescaled with `-r` for visualization , as you suggested in [issue #4](https://github.com/e-lab/ENet-training/issues/4). 

@ramonss Thank you for sharing your options. It was helpful. I was wondering whether this has anything to do with GPU amount until I saw yours. It is said some group has successfully reproduced it in MXNET, even with a `batchsize = 1`. But what could be the significant differences when we are training ENet based on different platforms?

**Comment by chuanzihe** [comment]
@codeAC29 It seems the results trained by @ramonss and me have the same problem in the lower part of image. Looks like the training was incomplete, which is contradict to the converged testing and training error. Any advice? Thank you again for answering so many of my issues. (because your work is truly brilliant!)

![image](https://cloud.githubusercontent.com/assets/19618625/19347988/36fcf20a-917e-11e6-9a9b-cb17719a3829.png)

**Comment by codeAC29** [comment]
The default option set in `opts.lua` were the options used by us. We never went beyond 300 epochs and generally network converged for us within ~250 epochs. Moreover, unlike the result which you are getting in the above image, detection of `ego vehicle` is crisp for us as you can see in the results provided in the paper.

**Comment by ramonss** [comment]
@codeAC29 Thanks for the answer. I tried again with the default options as follows below, but still no luck. The error does not go lower than 0.37 for encoder after 200 epochs and I have the same problem for the decoder part (errors are even higher due to accumulated errors from encoder).

th run.lua --dataset cs --datapath data/Cityscapes --model models/encoder.lua --save trained_models/encoder/ --cachepath /home/caduser/ENet-training/train/dataset_cache/encoder/ --imHeight 256 --imWidth 512 --labelHeight 32 --labelWidth 64

th run.lua --dataset cs --datapath data/Cityscapes --model models/decoder.lua --save trained_models/decoder/ --CNNEncoder trained_models/encoder/model-best.net --imHeight 256 --imWidth 512 --labelHeight 256 --labelWidth 512 --cachepath /home/caduser/ENet-training/train/dataset_cache/decoder/

In addition to the training problem of not converging, also when I try to use the trained model as you made available in https://www.dropbox.com/sh/dywzk3gyb12hpe5/AAD5YkUa8XgMpHs2gCRgmCVCa, I get the same problem in the visualization near the ego vehicle for Cityscapes. Perhaps I'm using the wrong command line for visualization? Could you please let me know if the following is correct?

qlua demo.lua -d ../train/trained_models/ -i ../train/data/Cityscapes/leftImg8bit/test/bonn/ -m github_sample -v --net 1

(where github_sample is the name of the folder where the trained model is located and 1 is the model number after I renamed model-best.net to model-1.net; that was the way I found to be able to run it without parameters errors in demo.lua)

I appreciate your help in advance.

**Comment by codeAC29** [comment]
Try the following two commands and let me know what do you get for them:

```
qlua demo.lua -d ../train/trained_models/ -i ../train/data/Cityscapes/leftImg8bit/test/bonn/ -m github_sample -v --net 1 -r 0.5
```

and

```
qlua demo.lua -d ../train/trained_models/ -i ../train/data/Cityscapes/leftImg8bit/test/bonn/ -m github_sample -v --net 1 -r 0.25
```

**Comment by ramonss** [comment]
@codeAC29 Thanks for the suggestion. It worked ;-) 

I can also confirm that even with high error values in the training (i.e. around 0.6 for decoder after 300 epochs), I can now see better results with the visualizer when resizing with 0.5 or 0.25. I guess that is because we are training with a resized version (512 x 256) of original Cityscapes images, isn't it?

Thanks again for your help. Really appreciated.

**Comment by codeAC29** [comment]
That is correct @ramonss . Since training was done on smaller resolution, it makes more sense (at least to me) to visualize it on smaller resolution.

**Comment by apaszke** [comment]
In fact, some of the default values in the repo didn't match the ones we reported in the paper (and the ones I've used to train on cityscapes). We used an lr of `5e-4` and wd of `2e-4`. Also we trained on 1024x512 images on Cityscapes. You might want to check these settings. Sorry for the trouble!

**Comment by adroit91** [comment]
Hi @apaszke

Any recommendations on different lr, wd, image dimensions to achieve same results as yours in the paper using this same code for the camVid dataset? We have tried light changes, but cannot go below 0.54 error, and even that converges around 35-45 epochs for encoder.

**Comment by TimoSaemann** [comment]
@clairehe @bajsk If you are still interested in a Caffe implementation, you may find this [ENet](https://github.com/TimoSaemann/ENet) repository helpful.
